### PR TITLE
Ryder Mint V2 on Stacks

### PR DESCRIPTION
**Pull request by MarvinJanssen**
You can try it in the console:

1. Add the new contract to Ryder NFT as a minter.
2. Enable it by calling `set-mint-enabled`.
3. Mint some to the new contract by calling `mint-to-contract`. The buffer length defines how many to mint.
4. Call `buy` to buy some.
5. Advance the chain by a few blocks.
6. Call `claim` to receive bought NFTs in random order.

### CODE REVIEW

File: backend-stacks/contracts/ryder-buy.clar [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    (try! (check-is-admin))
+    (asserts! (not (is-eq tx-sender new-admin)) err-not-allowed)
+    (ok (map-set admins new-admin value))))
+
+(define-private (burn-iter (token-id uint) (previous (response bool uint)))
+	(begin
+		(try! previous)
+		(as-contract (contract-call? .ryder-nft burn token-id))))
+
+(define-public (burn (token-ids (list 200 uint)))
```
Comment (friedger) [inline backend-stacks/contracts/ryder-buy.clar:70]: better name `burn-contract-tokens`

File: backend-stacks/contracts/ryder-buy.clar [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+(define-map admins principal bool)
+(map-set admins tx-sender true)
+
+(define-map token-mapping uint uint)
+
+(define-private (mint-self-iter (c (buff 1)) (p (response bool uint)))
+	(contract-call? .ryder-nft mint contract-principal))
+
+(define-public (mint-self (iterations (buff 200)))
```
Comment (friedger) [inline backend-stacks/contracts/ryder-buy.clar:32]: better name `mint-to-contract`

File: backend-stacks/contracts/ryder-buy.clar [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	(map-get? tier-prices (contract-call? .ryder-nft get-tier-by-token-id token-id)))
+
+(define-read-only (is-for-sale (token-id uint))
+	(is-eq (contract-call? .ryder-nft get-owner token-id) (ok (some contract-principal))))
+
+(define-public (buy (token-id uint))
+	(let ((price (unwrap! (get-token-price token-id) err-no-price)))
+		(unwrap! (stx-transfer? price tx-sender (var-get payment-recipient)) err-insufficient-balance)
+		(asserts! (var-get buy-enabled) err-mint-not-live)
+		(contract-call? .ryder-nft transfer token-id contract-principal tx-sender)))
```
Comment (friedger) [inline backend-stacks/contracts/ryder-buy.clar:48]: transfer should be wrapped in `as-contract`

**Review by friedger — COMMENTED** [review]
Looks good. 
Needs more testing.

File: backend-stacks/contracts/ryder-buy.clar (hunk 3 above)
Comment (MarvinJanssen) [reply backend-stacks/contracts/ryder-buy.clar:48]: No need. Check `transfer` in Ryder NFT.

File: backend-stacks/contracts/ryder-buy.clar (hunk 2 above)
Comment (MarvinJanssen) [reply backend-stacks/contracts/ryder-buy.clar:32]: Ok.

File: backend-stacks/contracts/ryder-buy.clar (hunk 1 above)
Comment (MarvinJanssen) [reply backend-stacks/contracts/ryder-buy.clar:70]: Ok.

**Comment by MarvinJanssen** [comment]
@friedger per discussion, `ryder-mint-v2` now follows a two-stage mechanism:

1. Call `buy` with the amount of NFTs you want to buy. It locks in a `block-height + const` block-height to claim the NFTs. The function transfers STX and reserves that number of tokens for the buyer.
2. Call `claim` at the previous `block-height + const` height to transfer NFTs from the contract to the buyer. Anyone can call `claim` for any buyer and block height.

**Comment by friedger** [comment]
claim-two, -five, -ten should be replaced with claim-many, because we have `buy` with an arbitrary number.

File: backend-stacks/contracts/ryder-buy.clar [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+(define-read-only (get-tier-price (tier-id uint))
+	(map-get? tier-prices tier-id))
+
+(define-read-only (get-token-price (token-id uint))
+	(map-get? tier-prices (contract-call? .ryder-nft get-tier-by-token-id token-id)))
+
+(define-read-only (is-for-sale (token-id uint))
+	(is-eq (contract-call? .ryder-nft get-owner token-id) (ok (some contract-principal))))
+
+(define-public (buy (token-id uint))
```
Comment (friedger) [inline backend-stacks/contracts/ryder-buy.clar:44]: Better `pre-order` ?

Comment (MarvinJanssen) [reply backend-stacks/contracts/ryder-buy.clar:44]: We can remove `ryder-buy` since we won't be going down that route for now.

**Comment by MarvinJanssen** [comment]
> claim-two, -five, -ten should be replaced with claim-many, because we have `buy` with an arbitrary number.

How would you implement it given the claim logic?

**Review by friedger — APPROVED** [review]

**Pull request merged**